example/bwmf/bwmf_task.go

```diff
@@ -168,19 +288,68 @@ func (t *bwmfTask) doDataReady(ctx context.Context, fromID uint64, method string
 
 // These two function carry out actual optimization.
 func (t *bwmfTask) updateDShard(ctx context.Context) {
-	select {
-	case <-time.After(time.Duration(rand.Intn(3)) * time.Second):
-	case <-ctx.Done():
-		return
+
+	if t.dims.N == -1 {
+		t.dims.N = 0
+		for _, m := range t.peerShards {
+			t.dims.N += len(m.Row)
+		}
+		wh := make([][]float32, t.dims.N)
+		for i, _ := range wh {
+			wh[i] = make([]float32, t.dims.m)
+		}
+		t.dLoss = &KLDivLoss{
+			V:  t.rowShard,
+			WH: wh,
+			m:  t.dims.N,
+			n:  t.dims.m,
+			k:  t.dims.k,
+		}
+	}
+
+	t.dLoss.W = NewBlocksParameter(&t.peerShards)
+	param := NewSingleBlockParameter(t.dShard)
+
+	ok := t.optimizer.Minimize(t.dLoss, t.stopCriteria, param)
```

Comment:
If any step is compute intensive, can you put a comment to mark them. This is quite important because we might need to use "ctx.Done" channel to abort repeated work.